lib/SILOptimizer/Utils/Existential.cpp

```diff
@@ -213,6 +213,15 @@ OpenedArchetypeInfo::OpenedArchetypeInfo(Operand &use) {
             getAddressOfStackInit(instance, user, isOpenedValueCopied)) {
       openedVal = stackInitVal;
     }
+
+    // Handle:
+    //   %4 = open_existential_ref
+    //   %5 = alloc_stack
+    //   store %4 to %5
+    auto firstUse = *instance->getUses().begin();
+    if (auto *store = dyn_cast<StoreInst>(firstUse->getUser())) {
+      openedVal = store->getSrc();
+    }
```

Comment:
The existential value stored on the stack here might not be the same value on the stack when it is opened.

Also, the value stored on the stack might no longer be live at the apply site.

There's a whole bunch of complexity in `getAddressOfStackInit` and `canReplaceCopiedValue` to deal with on-stack existentials.

Reply:
Hmm, I think I understand your first comment but, could you elaborate on the following?
> Also, the value stored on the stack might no longer be live at the apply site.

I've updated line 222 to use `getAddressOfStackInit(..., firstUse->getUser(), ...)` instead of a `dyn_cast<StoreInst>(...)`. Do you think this will be sufficient?

Reply:
In the abstract, we want to avoid producing a sequence like
```
r = open_existential_ref
destroy(r)
apply(r)
```
That shouldn't actually be a problem though because we're not in OSSA form so there will only be 'retain/release_value(r)' operations. As long as none of the retains get eliminated during this transformation, it should be safe.

A lot of the complexity in the existing code comes from the fact that it's handling 'open_existential_addr', which *can* be destroyed with 'destroy_addr'.

The main thing here is to prove that the stored value is the value used by the apply. One way to do that is to recognize all uses of the alloc_stack and make sure only one of them is a store. getStackInitInst already does that, but only handles copy_addr, not store.

Please try to write some .sil test cases to for corner cases.

Reply:
Thank you, that was extremely helpful. 

> A lot of the complexity in the existing code comes from the fact that it's handling 'open_existential_addr', which can be destroyed with 'destroy_addr'.

The constructor I modified does handle `open_existential_addr` but the specific optimization  I'm working on should only handle `open_existential_ref`, right?

> The main thing here is to prove that the stored value is the value used by the apply. One way to do that is to recognize all uses of the alloc_stack and make sure only one of them is a store. 

I've added more checking around the above code but, I suspect it isn't quite right yet. Should the stack alloc instruction ever be used in more than the initial store, the apply, and then the dealloc instructions? If not it may make it a bit easier to filter out what we can optimize. But that might be too harsh of a filter. 

Anyway, I've added many more tests and have uncovered several bugs. Right now I'm in the process of testing those bugs against master to see if they're my issue and if so, fixing them. I'm going to be pretty busy this week with other things so, turn around on this patch might be a bit slow.

Reply:
@atrick I've added a dominance order check which should ensure that there aren't those kinds of issues. I've also added one .sil test. I plan on writing some more, what other edge cases should I cover?

Reply:
If you want to generalize this pattern match, I think it should be part of the existing utilities so there's no new code in this function. `getStackInitInst` should be extended so it can return a store. It already checks for dominance of `ASIUser`. It's a bit more general than your new code. Whether it should set `isCopied` in that case is very subtle. The purpose of that flag is to make sure the original value isn't destroyed before the apply that needs to use it. In the case of a store it's simpler of we do not set `isCopied`, but then we should assert that the store's ownership qualifier is Unqualified (non-OSSA) and  comment when we support OSSA here, we need to insert a new copy of the value before the store (and make sure that copy is destroyed when replacing the apply operand).

FYI: in non-OSSA it should just work. If we have
```
retain val
store val to address
destroy val
use address
destroy_addr
```
It's actually ok in non-OSSA to reuse the original value like this:
```
retain val
store val to address
destroy val
use val
destroy_addr
```

Reply:
Good idea-- that's much cleaner. Also, thanks for the information about OSSA/non-OSSA and when copies are needed. Very interesting.